clang/test/SemaTemplate/recovery-crash-cxx20.cpp

```diff
@@ -0,0 +1,32 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c++20 %s
+
+namespace GH49093 {
+  class B {
+  public:
+    static int a() { return 0; } // expected-note {{member is declared here}}
+    decltype(a< 0 >(0)) test;    // expected-error {{member 'a' used before its declaration}}
```

Comment:
These diagnostics make me think there's a different bug that we're glossing over. 1) `a` is not being used before it's declaration; the declaration is right on line 6. 2) Why does this not get an "'a' is not a template" kind of diagnostic?

Reply:
Yeah, doing any amount of diagnostics before forming the unresolved lookup seems wrong.  I suspect the approach here is trying to bail out too quickly.

Reply:
@erichkeane @AaronBallman Thanks for the feedback. @AaronBallman The recovery checker attempts to diagnose an empty lookup and also tries to handle dependent member lookups, which detect them as being used before declarations

https://github.com/llvm/llvm-project/blob/df935ff4eca91013553edbf3ca1d45b568f6709f/clang/lib/Sema/SemaExpr.cpp#L2473-L2490

https://github.com/llvm/llvm-project/blob/df935ff4eca91013553edbf3ca1d45b568f6709f/clang/lib/Sema/SemaExpr.cpp#L2425

Is it acceptable to handle a lookup result that isn’t a template but uses explicit templated arguments at this stage?

Reply:
I think it generally DOES make sense to try the lookup?  We shouldn't crash of course, but failing to do the lookup pretending it is NOT a template makes a lot of sense to me.

Reply:
> I think it generally DOES make sense to try the lookup?

In the case of viable functions, the checker tries to recover even when the candidate list is empty. 

https://github.com/llvm/llvm-project/blob/416c3ce0138ff4039dab13ff634ee6392b9a3c7b/clang/lib/Sema/SemaOverload.cpp#L14056-L14059


https://github.com/llvm/llvm-project/blob/416c3ce0138ff4039dab13ff634ee6392b9a3c7b/clang/lib/Sema/SemaOverload.cpp#L13844-L13857

Once the lookup is complete

https://github.com/llvm/llvm-project/blob/df935ff4eca91013553edbf3ca1d45b568f6709f/clang/lib/Sema/SemaExpr.cpp#L2471

It identifies the expression as a dependent member

https://github.com/llvm/llvm-project/blob/df935ff4eca91013553edbf3ca1d45b568f6709f/clang/lib/Sema/SemaExpr.cpp#L2490

This is the entry point for these kinds of names, where we can try to stop the recovery process that causes the crash. Or should we skip building recovery call expression entirely for such cases?

Reply:
It sounds like we just need to 'try harder' to recover.  A fallback that treats it as a 'non template' would probably be better.  As far as the 'dependent member lookup', perhaps that needs to check for a recoveryexpr better and come up with a better recovery.

Reply:
I also thought skipping `DiagnoseDependentMemberLookup` and proceeding with the correction, as this seems to avoid those candidates and focus on the appropriate templated candidates

https://github.com/llvm/llvm-project/blob/3138eb500c9462bcb6899088d49644adb4d90f62/clang/lib/Sema/SemaExpr.cpp#L2500-L2612

since `DiagnoseDependentMemberLookup` treats these candidates as valid, it continues with recovery and constructs the call expression as entirely valid 

https://github.com/llvm/llvm-project/blob/9ac00b85e05d21be658d6aa0c91cbe05bb5dbde2/clang/lib/Sema/SemaOverload.cpp#L13899-L13901

https://github.com/llvm/llvm-project/blob/3138eb500c9462bcb6899088d49644adb4d90f62/clang/lib/Sema/SemaExpr.cpp#L6528-L6530

https://github.com/llvm/llvm-project/blob/9ac00b85e05d21be658d6aa0c91cbe05bb5dbde2/clang/lib/Sema/SemaOverload.cpp#L14145

https://github.com/llvm/llvm-project/blob/9ac00b85e05d21be658d6aa0c91cbe05bb5dbde2/clang/lib/Sema/SemaOverload.cpp#L13937

Reply:
I also thought skipping `DiagnoseDependentMemberLookup` and proceeding with the correction, as this seems to avoid those candidates and focus on the appropriate templated candidates

https://github.com/llvm/llvm-project/blob/3138eb500c9462bcb6899088d49644adb4d90f62/clang/lib/Sema/SemaExpr.cpp#L2500-L2612

since `DiagnoseDependentMemberLookup` treats these candidates as valid, it continues with recovery and constructs the call expression as entirely valid 

https://github.com/llvm/llvm-project/blob/9ac00b85e05d21be658d6aa0c91cbe05bb5dbde2/clang/lib/Sema/SemaOverload.cpp#L13899-L13901

https://github.com/llvm/llvm-project/blob/3138eb500c9462bcb6899088d49644adb4d90f62/clang/lib/Sema/SemaExpr.cpp#L6528-L6530

https://github.com/llvm/llvm-project/blob/9ac00b85e05d21be658d6aa0c91cbe05bb5dbde2/clang/lib/Sema/SemaOverload.cpp#L14145

https://github.com/llvm/llvm-project/blob/9ac00b85e05d21be658d6aa0c91cbe05bb5dbde2/clang/lib/Sema/SemaOverload.cpp#L13937

Reply:
> A fallback that treats it as a 'non template' would probably be better

@AaronBallman @erichkeane Do you mean that the checker can construct a recovery call expression by skipping the `ExplicitTemplateArgs` in such cases?

Reply:
I'm not explicitly saying that (hah!).  But the behavior you have here is not correct.  First, I still don't see why it would say 'used before its declaration', when its declaration is right above it.  Second, SHOULD say something like, "non template used with template args, what are you doing?" sorta error.  

As far as recovery, it could/should probably treat it as if you HADN'T used the template arguments and continue, so any recovery (for the type reference?) should be based on the found-but-not-template type.

Reply:
I added the changes and the [test](https://github.com/llvm/llvm-project/pull/107786/files#diff-035f4e94708ebca719a751e5e44a5a761610649239bc3cec2593a670929c8d01R27-R32) no longer throws the _`used before its declaration`_ error. From my understanding, if recovery detects explicit arguments, the checker should try looking up candidates with templates. If it finds a match, recovery can continue; if not, it should throw an error and attempt typo correction. Does that make sense?

Reply:
@AaronBallman @erichkeane could you kindly take a look at the latest changes?